src/blockMonitor/blockchainMachine.ts

```diff
@@ -31,62 +106,92 @@ export class BlockchainMachine<TBlock extends IBlockStub> extends StartStopServi
+        if (this.componentNames.has(component.name)) throw new ApplicationError(`A Component with the name "${component.name}" was already added.`);
 
-        this.componentsAndStates.push({
-            component,
-            states: new WeakMap()
-        });
+        this.components.push(component);
+        this.componentNames.add(component.name);
     }
 
-    private processNewBlock(block: TBlock) {
-        // Every time a new block is received we calculate the anchor state for that block and store it
+    private async processNewBlock(block: TBlock) {
+        try {
+            await this.lock.acquire();
 
-        for (const { component, states } of this.componentsAndStates) {
-            // If the parent is available and its anchor state is known, the state can be computed with the reducer.
-            // If the parent is available but its anchor state is not known, first compute its parent's initial state, then apply the reducer.
-            // Finally, if the parent is not available at all in the block cache, compute the initial state based on the current block.
+            // Every time a new block is received we calculate the anchor state for that block and store it
 
-            let newState: AnchorState;
-            if (this.blockProcessor.blockCache.hasBlock(block.parentHash)) {
-                const parentBlock = this.blockProcessor.blockCache.getBlock(block.parentHash);
-                const prevAnchorState = states.get(parentBlock) || component.reducer.getInitialState(parentBlock);
+            for (const component of this.components) {
+                // If the parent is available and its anchor state is known, the state can be computed with the reducer.
+                // If the parent is available but its anchor state is not known, first compute its parent's initial state, then apply the reducer.
+                // Finally, if the parent is not available at all in the block cache, compute the initial state based on the current block.
 
-                newState = component.reducer.reduce(prevAnchorState, block);
-            } else {
-                newState = component.reducer.getInitialState(block);
-            }
+                let newState: AnchorState;
+                let prevHeadAnchorState: AnchorState | null = null;
+                if (this.blockProcessor.blockCache.hasBlock(block.parentHash)) {
+                    const parentBlock = this.blockProcessor.blockCache.getBlock(block.parentHash);
 
-            states.set(block, newState);
+                    if (parentBlock) {
+                        prevHeadAnchorState = this.blockItemStore.getItem(parentBlock.hash, `${component.name}:prevEmittedState`);
+                    }
+
+                    const prevAnchorState =
+                        this.blockItemStore.getItem(parentBlock.hash, `${component.name}:state`) || component.reducer.getInitialState(parentBlock);
```

Comment:
All item keys like `state` and `attached` etc. should be public static readonly strings on the BlockItemStore

Reply:
Done.

I am not sure I like having these keys in the BlockItemStore (meaning that the BlockItemStore has to know about all the keys used by other classes), but that could be part of the improvements in #316.

Reply:
Keys now all in BlockItemStore and private.